src/dotty/tools/dotc/core/tasty/TreeUnpickler.scala

```diff
@@ -590,7 +590,7 @@ class TreeUnpickler(reader: TastyReader, tastyName: TastyName.Table) {
         case TYPEDEF | TYPEPARAM =>
           if (sym.isClass) {
             val companion = sym.scalacLinkedClass
-            if (companion != NoSymbol) {
+            if (companion != NoSymbol && symAtAddr.valuesIterator.contains(companion)) {
```

Comment:
`valuesIterator` takes linear time to check if element is contained.
It means that this change makes reading tasty quadratic in number of `TypeDefs`. 

if `companion != NoSymbol`, where does `companion` originate from if not from Tasty?